suite/tests/client-interface/mangle_suspend.dll.c

```diff
@@ -0,0 +1,190 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Test xl8 pc of rip-rel instruction (xref #3307) caused by
+ * asynch interrupt.
+ */
+
+#include "dr_api.h"
+#include "drmgr.h"
+#include "client_tools.h"
+#include "mangle_suspend-shared.h"
+#include <string.h> /* memset */
+
+#define MAX_RESUME_COUNT 10
+
+#define CHECK(x, msg)                                                                \
+    do {                                                                             \
+        if (!(x)) {                                                                  \
+            dr_fprintf(STDERR, "CHECK failed %s:%d: %s\n", __FILE__, __LINE__, msg); \
+            dr_abort();                                                              \
+        }                                                                            \
+    } while (0);
+
+extern void
+test_asm();
+
+static byte *add_instr_pc = NULL;
+
+static dr_emit_flags_t
+event_app_analysis(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                   bool translating, OUT void **user_data)
+{
+    instr_t *inst;
+    bool prev_was_mov_const = false;
+    ptr_int_t val1, val2;
+    *user_data = NULL;
+
+    if (translating)
+        return DR_EMIT_DEFAULT;
+
+#ifdef X86_64
+    /* Look for duplicate mov immediates telling us which subaction we're in */
+    for (inst = instrlist_first_app(bb); inst != NULL; inst = instr_get_next_app(inst)) {
+        if (instr_is_mov_constant(inst, prev_was_mov_const ? &val2 : &val1)) {
+            if (prev_was_mov_const && val1 == val2 &&
+                val1 != 0 && /* rule out xor w/ self */
+                opnd_is_reg(instr_get_dst(inst, 0)) &&
+                opnd_get_reg(instr_get_dst(inst, 0)) == SUSPEND_TEST_REG) {
+                *user_data = (void *)val1;
+                instrlist_meta_postinsert(bb, inst, INSTR_CREATE_label(drcontext));
+            } else
+                prev_was_mov_const = true;
+        } else
+            prev_was_mov_const = false;
+    }
+#else
+        /* XXX i#3307: port to ARM if possible. */
+#endif
+    return DR_EMIT_DEFAULT;
+}
+
+static void
+suspend_func()
+{
+#ifdef X86_64
+    void **drcontexts = NULL;
+    uint num_suspended;
+    if (dr_suspend_all_other_threads(&drcontexts, &num_suspended, NULL)) {
+        if (num_suspended != 1)
+            CHECK(false, "num_suspended unexpected!");
+        dr_mcontext_t mc = { sizeof(mc), DR_MC_INTEGER | DR_MC_CONTROL };
+        if (!dr_get_mcontext(drcontexts[0], &mc))
+            CHECK(false, "dr_get_mcontext failed!");
+        /* Check xip such that we only perform the check if we are in the loop
+         * body of the test, in order to avoid races.
+         */
+        if (mc.xip >= add_instr_pc - 7 && mc.xip <= add_instr_pc + 24) {
```

Comment:
Magic constants 7 and 24: please explain in a comment; a comment in the asm code would be good too to ensure these stay in synch.

Reply:
Done